src/stan/services/util/initialize.hpp

```diff
@@ -13,202 +13,211 @@
+                  gradient, &log_prob_msg);
+    } catch (const std::exception& e) {
+      if (log_prob_msg.str().length() > 0)
+        logger.info(log_prob_msg);
+      logger.info(e.what());
+      throw;
+    }
+    clock_t end_check = clock();
+    double deltaT = static_cast<double>(end_check - start_check)
+                    / CLOCKS_PER_SEC;
+    if (log_prob_msg.str().length() > 0)
+      logger.info(log_prob_msg);
 
-            logger.info("Adjust your expectations accordingly!");
-            logger.info("");
-            logger.info("");
-          }
-          if (gradient_ok)
-            break;
-        }
+    for (size_t i = 0; i < gradient.size(); ++i) {
+      if (gradient_ok && !boost::math::isfinite(gradient[i])) {
+        logger.info("Rejecting initial value:");
+        logger.info("  Gradient evaluated at the initial value"
+                    " is not finite.");
+        logger.info("  Stan can't start sampling from this"
+                    " initial value.");
+        gradient_ok = false;
+      }
+    }
+    if (gradient_ok && print_timing) {
+      logger.info("");
+      std::stringstream msg1;
+      msg1 << "Gradient evaluation took " << deltaT << " seconds";
+      logger.info(msg1);
 
-        if (num_init_tries == MAX_INIT_TRIES) {
-          if (!is_fully_initialized && !init_zero) {
-            logger.info("");
-            std::stringstream msg;
-            msg << "Initialization between (-" << init_radius
-                << ", " << init_radius << ") failed after"
-                << " " << MAX_INIT_TRIES <<  " attempts. ";
-            logger.info(msg);
-            logger.info(" Try specifying initial values,"
-                        " reducing ranges of constrained values,"
-                        " or reparameterizing the model.");
-          }
-          throw std::domain_error("Initialization failed.");
-        }
+      std::stringstream msg2;
+      msg2 << "1000 transitions using 10 leapfrog steps"
+           << " per transition would take"
+           << " " << 1e4 * deltaT << " seconds.";
+      logger.info(msg2);
 
-        init_writer(unconstrained);
-        return unconstrained;
-      }
+      logger.info("Adjust your expectations accordingly!");
+      logger.info("");
+      logger.info("");
+    }
+    if (gradient_ok)
+      break;
+  }
 
+  if (num_init_tries == MAX_INIT_TRIES) {
+    if (!is_fully_initialized && !init_zero) {
+      logger.info("");
+      std::stringstream msg;
+      msg << "Initialization between (-" << init_radius
+          << ", " << init_radius << ") failed after"
+          << " " << MAX_INIT_TRIES <<  " attempts. ";
+      logger.info(msg);
+      logger.info(" Try specifying initial values,"
+                  " reducing ranges of constrained values,"
+                  " or reparameterizing the model.");
     }
+    throw std::domain_error("Initialization failed.");
   }
+
+  init_writer(unconstrained);
```

Comment:
Does `init_writer` modify `unconstrained`?  I have no idea what this is doing here, so some doc may help.

Reply:
No, it does not. It just outputs the initialization. This is used for ADVI and it really is necessary if we want to think about restarting.